firebase-firestore/src/androidTest/java/com/google/firebase/firestore/QueryTest.java

```diff
@@ -483,6 +486,79 @@ public void testQueriesFireFromCacheWhenOffline() {
+    Map<String, Object> docA = map("zip", 98101L);
+    Map<String, Object> docB = map("zip", 91102L);
+    Map<String, Object> docC = map("zip", "98101");
+    Map<String, Object> docD = map("zip", asList(98101L));
+    Map<String, Object> docE = map("zip", asList("98101", map("zip", 98101L)));
+    Map<String, Object> docF = map("zip", map("code", 500L));
+    Map<String, Object> docG = map("zip", asList(98101L, 98102L));
+    Map<String, Object> docH = map("code", 500L);
+    Map<String, Object> docI = map("zip", null);
+    Map<String, Object> docJ = map("zip", Double.NaN);
+
+    Map<String, Map<String, Object>> allDocs =
+        map(
+            "a", docA, "b", docB, "c", docC, "d", docD, "e", docE, "f", docF, "g", docG, "h", docH,
+            "i", docI, "j", docJ);
+    CollectionReference collection = testCollectionWithDocs(allDocs);
+
+    // Search for zips not matching 98101.
+    Map<String, Map<String, Object>> expectedDocsMap = Maps.newHashMap(allDocs);
+    expectedDocsMap.remove("a");
+    expectedDocsMap.remove("h");
+    expectedDocsMap.remove("i");
+
+    QuerySnapshot snapshot = waitFor(collection.whereNotEqualTo("zip", 98101L).get());
+    assertSetEquals(expectedDocsMap.values(), querySnapshotToValues(snapshot));
+
+    // With objects.
+    expectedDocsMap = Maps.newHashMap(allDocs);
+    expectedDocsMap.remove("f");
+    expectedDocsMap.remove("h");
+    expectedDocsMap.remove("i");
+    snapshot = waitFor(collection.whereNotEqualTo("zip", map("code", 500)).get());
+    assertSetEquals(expectedDocsMap.values(), querySnapshotToValues(snapshot));
+
+    // With Null.
+    expectedDocsMap = Maps.newHashMap(allDocs);
+    expectedDocsMap.remove("h");
+    expectedDocsMap.remove("i");
+    snapshot = waitFor(collection.whereNotEqualTo("zip", null).get());
+    assertSetEquals(expectedDocsMap.values(), querySnapshotToValues(snapshot));
+
+    // With NaN.
+    expectedDocsMap = Maps.newHashMap(allDocs);
+    expectedDocsMap.remove("h");
+    expectedDocsMap.remove("i");
+    expectedDocsMap.remove("j");
+    snapshot = waitFor(collection.whereNotEqualTo("zip", Double.NaN).get());
+    assertSetEquals(expectedDocsMap.values(), querySnapshotToValues(snapshot));
+  }
+
+  // TODO(ne-queries): Re-enable once emulator support is added to CI.
+  @Ignore
+  @Test
+  public void testQueriesCanUseNotEqualFiltersWithDocIds() {
+    Map<String, String> docA = map("key", "aa");
+    Map<String, String> docB = map("key", "ab");
+    Map<String, String> docC = map("key", "ba");
+    Map<String, String> docD = map("key", "bb");
+    Map<String, Map<String, Object>> testDocs =
+        map(
+            "aa", docA,
+            "ab", docB,
+            "ba", docC,
+            "bb", docD);
+    CollectionReference collection = testCollectionWithDocs(testDocs);
+    QuerySnapshot docs = waitFor(collection.whereNotEqualTo(FieldPath.documentId(), "aa").get());
+    assertEquals(asList(docB, docC, docD), querySnapshotToValues(docs));
```

Comment:
Apologies for what's going to be an irritating question (since I didn't pick up on this earlier), but why do you `assertSetEquals(x, expected)` in the test above but `assertEquals(asList(x), expected)` here when we're testing the same underlying operator?

To me it seems like either the ordering matters or it doesn't. If ordering of query results is predictable, `assertSetEquals` seems wrong because it allows the result to be unordered when our customers really care that it's ordered. The reverse is also true: if ordering *doesn't* matter, then this test is too strict, and could create a false positive.

To my understanding, query results always have a well defined ordering (breaking ties on key ordering). This seems to point to `assertSetEquals` being wrong above.

What's your take on this? Why the discrepancy?

(The same concern applies to the `whereNotIn`, below).

Reply:
Thanks for catching this actually! I originally ran into an issue where I was converting the snapshots into a HashSet in order to easily remove the unneeded documents, but converting them back into a List resulted in an ordering that didn't match the query snapshots from the query. Instead of trying to create the `expectedDocsMap` in the correct order, I took the "shortcut" of comparing them as sets to avoid the ordering errors without realizing that order matters in query results.

I took a look at the code, and realized that Using `Maps.newHashMap()` keeps the documents in the correct order. Switched back to using `assertEquals` with `Lists.newArrayList()`.